### PR TITLE
refactor: Avoid v_array copies

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/cb_explore_adf_common.h [hunk 1]
```diff
@@ -123,16 +123,16 @@ inline void cb_explore_adf_base<ExploreType>::predict(
...
+    label_example->l.cb = std::move(data._empty_label);
   }
 
   data.explore.predict(base, examples);
 
   if (label_example != nullptr)
   {
     // predict path, restore label
-    label_example->l.cb = data._action_label;
+    label_example->l.cb = std::move(data._action_label);
```
Comment (jackgerrits) [inline vowpalwabbit/cb_explore_adf_common.h:135]: Shouldn't `data._empty_label` be moved back into down here?

Comment (bassmang) [reply vowpalwabbit/cb_explore_adf_common.h:136]: Just changed this in the last commit

**Review by olgavrou — APPROVED** [review]

File: vowpalwabbit/cb_explore_adf_common.h (hunk 1 above)
Comment (olgavrou) [reply vowpalwabbit/cb_explore_adf_common.h:136]: @jackgerrits should we also reset the label here? i.e. clear the array and reset the cost?

Comment (jackgerrits) [reply vowpalwabbit/cb_explore_adf_common.h:136]: Yeah we probably should

File: vowpalwabbit/cb_explore_adf_common.h [hunk 2]
```diff
@@ -123,16 +123,18 @@ inline void cb_explore_adf_base<ExploreType>::predict(
...
+    data._empty_label.costs.delete_v();
```
Comment (peterychang) [inline vowpalwabbit/cb_explore_adf_common.h:136]: Do we need to delete the v_array? Can we just clear it instead?

File: vowpalwabbit/explore_eval.cc [hunk 3]
```diff
@@ -127,13 +127,15 @@ void do_actual_learning(explore_eval& data, multi_learner& base, multi_ex& ec_se
...
-    label_example->l.cb = data.empty_label;
+    data.action_label = std::move(label_example->l.cb);
+    label_example->l.cb = std::move(data.empty_label);
   }
   multiline_learn_or_predict<false>(base, ec_seq, data.offset);
 
   if (label_example != nullptr)  // restore label
-    label_example->l.cb = data.action_label;
+    label_example->l.cb = std::move(data.action_label);
+    data.empty_label.costs.delete_v();
```
Comment (jackgerrits) [inline vowpalwabbit/explore_eval.cc:138]: `delete_v` should no longer be used. Do you mean `clear` instead?

File: vowpalwabbit/cb_explore_adf_common.h (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/cb_explore_adf_common.h]: Just recommitted with clear

File: vowpalwabbit/explore_eval.cc (hunk 3 above)
Comment (bassmang) [reply vowpalwabbit/explore_eval.cc]: Just recommitted with clear

**Review by olgavrou — APPROVED** [review]

**Pull request merged**